WordPress/gutenberg
# Fix: duplicate markup on anchor attribute

Fixes: #48232
Fixes: #49382
Fixes: #49578

Related to: #44771
Alternative to #48246

## What?

This PR fixes a problem that `id` attribute is saved to both the comment delimiter and markup in blocks that support `anchor` attribute. This will prevent block validation from breaking blocks that previously stored the `id` attribute only in the markup.

## Why?

In #44771, dynamic blocks now support the `anchor`. The server registers the `anchor` attribute for all blocks that support the `anchor` in the `register_attributes` callback function. The attribute then has [only a `type` property of type `string`](https://github.com/WordPress/gutenberg/blob/c846c62eabcafc1d61973bc5fe9ac43b07e85365/lib/block-supports/anchor.php#L24-L26).

However, previous blocks that supported the `anchor` [read the `id` attribute from the markup](https://github.com/WordPress/gutenberg/blob/c846c62eabcafc1d61973bc5fe9ac43b07e85365/packages/block-editor/src/hooks/anchor.js#L23-L28) and don't store the value in the comment delimiter. If the `anchor` attribute is already registered on the server side, it will [take precedence](https://github.com/WordPress/gutenberg/blob/c846c62eabcafc1d61973bc5fe9ac43b07e85365/packages/block-editor/src/hooks/anchor.js#L39-L42), so its definition will be different from the conventional attributes.

As a result, I think the difference in the definition of this property caused the generation of incorrect markup, the `anchor` attribute as an unnecessary comment delimiter, and the failure of block validation as reported in #48232.

I think we need to separate the definition of the `anchor` property between blocks that save the `id` attribute as markup and those that don't save any markup in the `save` function, as many dynamic blocks do.

## How?

On the server side, we don't know whether the block saves the `id` attribute as markup, i.e., whether the `save` function returns null or not. ~~Therefore, I have decided NOT to register the `anchor` attribute on the server side, but to check the rendering result of the `save` function on the client side and register the `anchor` attribute with an appropriate definition according to the result.~~

## Testing Instructions

~~The addition of server-side anchor support has already been incorporated into the core, so comment out the following line:~~

~~https://github.com/WordPress/wordpress-develop/blob/7d8b5e89bc8b980e74eb7ba39b10ba68d52da7b4/src/wp-includes/block-supports/anchor.php#L70~~
(@edit: In #48592, anchor support in core blocks has been reverted. Therefore, the above is not necessary.)

Confirm that anchor attributes are saved and rendered as expected in the following two types of blocks.

### ~~Blocks that have a `save` function~~
### Blocks with anchor property `true` or `"html"`

- As before, the `id` attribute should be saved as part of the markup and not in the comment delimiter.

### ~~Blocks that don't have a `save` function or return null~~
### Blocks with anchor property `delimiter`

- The `anchor` attribute should be saved in the comment delimiter.
- Blocks rendered server-side by the `ServerSideRender` component should not cause errors when the `anchor` attribute is set.
